doc/automaatiotestit.md

```diff
@@ -0,0 +1,215 @@
+Testiluokat saattavat kuitenkin myös sisältää datan alustukseen, tilanteiden tarkistukseen tai vastaavaan muuhun
+testeihin liittyvään toiminnallisuuteen liittyviä funktioita, jotka eivät sisällä `@Test`-annotaatiota, sillä ne eivät
+ole varsinaisia testejä, joita tulisi automaattisesti suorittaa.
+
+Kotlin tukee testifunktioiden nimeämistä välilyöntejä hyödyntäen, joka olisi suositeltu tapa:
+
+```
+@Test
+fun äläNimeäTestiäNäin()
+
+@Test
+fun `Nimeä testi mielummin näin`()
+```
+
+# Yksikkötestit (Unit tests)
+
+Yksikkötestillä eli unit-testillä tarkoitetaan yksinkertaista, yhdessä prosessissa suoritettavaa, rajattua
+toiminnallisuutta testaavaa testiä. Tyypillinen yksikkötesti kokeilee, että yksi funktio toimii oikealla tavalla.
+Esimerkki yksikkötestistä olisi vaikkapa varmistus, että jotain oliota formatoiva funktio palauttaa oletetun
+merkkijonon.
+
+Yksikkötestien kirjoittaminen helpottuu, kun varsinainen testattava logiikka on kirjoitettu puhtaana funktiona. Lyhyesti
+kuvattuna puhtaat funktiot ovat sellaisia, joiden arvo perustuu pelkästään funktiolle annettuihin argumentteihin ja joka
+palauttaa tismalleen identtisen arvon samoilla argumenteilla, riippumatta ohjelmiston muusta tilasta.
+
+Koska yksikkötestit täytyy olla mahdollista suorittaa nopeasti yhdessä prosessissa (säikeessä), ne eivät
+voi esimerkiksi ottaa tietokantaan yhteyksiä. Yksikkötestejä hyödynnetään toistaiseksi lähinnä Kotlin-koodin puolella.
+
+## Yksikkötestien sijainti ja nimitys
+
+Geoviitteen yksikkötestit sijaitsevat tietylle toiminnallisuudelle varatussa tiedostohierarkiassa, ja testikoodin
+polkurakenne vastaa varsinaisen logiikan polkurakennetta. Kun varsinainen logiikka löytyy esimerkiksi polusta
+
+```
+infra/src/main/kotlin/fi/fta/geoviite/infra/geometry
+```
+
+niin geometry-pakettiin liittyvä yksikkötestikoodi löytyy vastaavasti polusta
+
+```
+infra/src/test/kotlin/fi/fta/geoviite/infra/geometry
+```
+
+joiden ainut ero on kansio 'main|test' polun alkupuolella.
+
+Yksikkötestitiedostojen nimet noudattavat rakennetta `<TestattavaOminaisuusTiedosto>Test.kt`, eli tiedoston nimi alkaa
+main-puolelta löytyvällä tiedostonimellä. Vastaavasti yksikkötestitiedosto taas päättyy `Test`-loppuliitteellä.
+
+Siispä esimerkiksi varsinaista logiikkaa sisältävän tiedoston
+
+```
+infra/src/main/kotlin/fi/fta/geoviite/infra/geometry/ElementListing.kt
+```
+
+yksikkötestit löytyvät tiedostosta
+
+```
+infra/src/test/kotlin/fi/fta/geoviite/infra/geometry/ElementListingTest.kt
+```
+
+# Integraatiotestit (integration tests)
+
+Integraatiotestit voivat hyödyntää useampaa kuin yhtä prosessia (säiettä). Tämä tarkoittaa, että ne voivat hyödyntää
+esimerkiksi tietokantaa taikka ulkopuolisia palveluita. Geoviitteen integraatiotestit tarkoittavat käytännössä kuitenkin
+oikeaa tietokantaprosessia hyödyntäviä testejä. Geoviitteen integraatiotestit eivät siis lähtökohtaisesti käytä
```

Comment:
Näen itse yksikkö- ja integraatiotestien välisen fundamentaalisen eron enemmän erona testin tarkoitusperässä ja (paremman termin puutteessa) "syvyydessä," joten tämä säikeiden kautta asiaa tarkastelu kinnaa jotenkin mulla vastakarvaan. Nään itse yksikkötestit testeinä, jotka testaavat mahdollisimman pienen ja mistään muusta riippumattoman kvantin (ns. "yksikön",) ja integraatiotestit testeinä, jotka testaavat jotain kokonaisuutta ja _integroivat_ testiin mukaan muutakin kokonaisuuden testaamiseen tarvittavaa tavaraa, esimerkiksi meidän testien tapauksessa tietokannan tai jonkun muun ulkoisen palikan (esim. Ratkon tai Projektivelhon mock-rajapinnan muodossa.) Täten esimerkiksi funktiolle, jonka tarkoitus on rinnakkaistaa joku operaatio, voi silti fundamentaalisesti mielestäni kirjoittaa yksikkötestejä ihan siinä missä yksisäikeisillekin ja ilmaisisin tämän eroavaisuuden jotenkin muuten.

Reply:
Säikeiden/prosessien kautta tarkastelussa on mun kokemuksella ainakin se hyöty, että se antaa melko eksplisiittisen rajauksen kompleksisuudelle (kunhan tietää mitä prosessi/säie tarkoittavat). Tuosta sunkin kuvauksesta on aika vaikeaa ymmärtää mitä määritelmällistä eroa tarkoitat yksikkö- ja integraatiotestin välillä (kuvauksesi on aika abstrakti), ja vielä vaikeampaa jollei näihin liittyen ole kokemuspohjaa jo ennestään. Sen sijaan varsinaista tulkintaeroa en edes huomannut, ihan vastaavasti tässäkin on tuota syvyyttä pyritty kuvaamaan, ja kompleksisuuteen on yhdistetty tuo prosessi/säiemäärä jolle voi myös mieltää lukumäärän. Minusta intuitiivisesti kuvaava rajaus (grounding), mutta ei tietysti kaikille.

Tämänhän pyrkimys onkin olla ns. pehmeä määritelmä tätä koodikantaa varten, sillä terminologisia ja tulkinnallisia eroja on jokaisessa koodikannassa miten näitä käytetään. Se mikä olisi Geoviitteen kannalta hyödyllinen määritelmä kuuluisi tähän, mutta siihen pääseminen saattaa kestää hetken.